augur/export_v2.py

```diff
@@ -1144,19 +1146,22 @@ def run(args):
 
     # Write outputs - the (unified) dataset JSON intended for auspice & perhaps the ref root-sequence JSON
     indent = {"indent": None} if args.minify_json else {}
-    write_json(data=orderKeys(data_json), file_name=args.output, include_version=False, **indent)
-
-    if args.include_root_sequence:
+    if args.include_root_sequence or args.include_root_sequence_inline:
+        # Note - argparse enforces that only one of these args will be true
         if 'reference' in node_data:
-            # Save the root sequence with the same stem from the main auspice
-            # output filename.  For example, if the main auspice output is
-            # "auspice/zika.json", the root sequence will be
-            # "auspice/zika_root-sequence.json".
-            output_path = Path(args.output)
-            root_sequence_path = output_path.parent / Path(output_path.stem + "_root-sequence" + output_path.suffix)
-            write_json(data=node_data['reference'], file_name=root_sequence_path, include_version=False, **indent)
+            if args.include_root_sequence_inline:
+                data_json['root_sequence'] = node_data['reference']
+            else:
```

Comment:
Even though "argparse enforces that only one of these args will be true", it'd be better practice to also be explicit and comprehensive in this conditional, e.g. 

```python
if args.include_root_sequence_inline:
    ...
elif args.include_root_sequence:
    ...
else:
   raise AssertionError("…")
```

argparse is "far away" from this code and it'd be easy for things to get out of sync.

Reply:
So this is an easy change, and I won't stand in its way, but how should one think about this more generally? E.g. if a command accepted vcf or fasta inputs, should we always write `if (vcf) ... elif (fasta) ... else raise`, even once we've validated that there's only one present? Or is this a judgement call related to how "far away" the checking is? I see this as introducing unreachable code, and would expect a good linter would flag it as such.

Reply:
> So this is an easy change, and I won't stand in its way, but how should one think about this more generally?

If you'd rather, I would also think it fine to do:

```python
if args.include_root_sequence_inline:
    ...
elif args.include_root_sequence:
    ...
```

The failure mode is then limited to not doing anything if we change the options/enclosing conditional instead of erroring, but at least the wrong thing is still not done.

My preference is not for the `else` or error specifically, but to positively explicitly match enumerated cases instead of implicitly matching one as a fallback.

> E.g. if a command accepted vcf or fasta inputs, should we always write `if (vcf) ... elif (fasta) ... else raise`, even once we've validated that there's only one present?

I would say so, yes!  Or at least use if/elif, not just if/else.

> Or is this a judgement call related to how "far away" the checking is? I see this as introducing unreachable code, and would expect a good linter would flag it as such.

There's some element of a judgment call, e.g. "crosses a function boundary" seems definitely far away enough to me but "within the same function" is more blurry.  I see this as not as unreachable code but akin to asserting an invariant the code requires, much like how `assert` statements are used (which we also expect to be "unreachable" in all but very exceptional cases, and when they are its usually a programming error).

Reply:
Changed to if/elif, which seems reasonable given the enclosing conditional ensuring at least one is true is ~3 lines earlier.

P.S. the hypothetical about `if (vcf) ... else ... ` [is](https://github.com/nextstrain/augur/blob/master/augur/translate.py#L428-L433) [not](https://github.com/nextstrain/augur/blob/master/augur/ancestral.py#L118-L128) [so](https://github.com/nextstrain/augur/blob/master/augur/translate.py#L376-L384) [hypothetical](https://github.com/nextstrain/augur/blob/master/augur/translate.py#L441-L456)

Reply:
> P.S. the hypothetical about …

Yep, I'm aware!

Reply:
> given the enclosing conditional ensuring at least one is true is ~3 lines earlier.

I realize it's 3 lines earlier and totally see how on the surface it might seem silly/extraneous to add additional checks here and now, but refactoring down the road could easily move these things (much) further apart. Such constructions are common sources of bugs IME, and one that's easy to remove from possibility by the small practice change of always enumerating cases.